api/src/main/java/org/apache/iceberg/io/LocationProvider.java

```diff
@@ -28,6 +28,19 @@
  * <p>Implementations must be {@link Serializable} because instances will be serialized to tasks.
  */
 public interface LocationProvider extends Serializable {
+
+  /**
+   * Returns the fully-qualified data location for a table, where data files such as Parquet, ORC,
+   * or Avro are stored. The table data location can be retrieved from the table properties input,
+   * such as 'write.data.path', or from 'write.object-storage.path' (deprecated), or from
+   * 'write.folder-storage.path' (deprecated). Refer to the implementation class of {@link
+   * LocationProvider} for more details.
+   */
+  default String dataLocation() {
+    throw new UnsupportedOperationException(
+        String.format("%s does not expose dataLocation", this.getClass().getCanonicalName()));
```

Comment:
This breaks old location providers, they just would throw UnsupportedOperationException instead of MethodNotFound exceptions.

If we want the method to be backwards compatible it must return a signal instead (like null) and the rest of the code should treat that as iff this method was not implemented.

Reply:
hi @RussellSpitzer As discussed, I removed dataLocation() new API from this PR and will add it as separate PR